clippy_lints/src/option_manual_map.rs

```diff
@@ -0,0 +1,200 @@
+                        format!(
+                            "|{}| {}",
+                            some_binding,
+                            snippet_with_applicability(cx, some_expr.span, "..", &mut app),
+                        )
+                    }
+                };
+
+                span_lint_and_sugg(
+                    cx,
+                    OPTION_MANUAL_MAP,
+                    expr.span,
+                    "manual implementation of `Option::map`",
+                    "use `map` instead",
+                    format!("{}.map({})", scrutinee, body),
+                    app,
+                );
+            }
+        }
+    }
+}
+
+enum SomeBinding {
+    Wild,
+    Name(Ident),
+}
+impl fmt::Display for SomeBinding {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Wild => f.write_str("_"),
+            Self::Name(name) => name.fmt(f),
+        }
+    }
+}
+
+fn get_some_expr(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> Option<&'tcx Expr<'tcx>> {
+    match expr.kind {
+        ExprKind::Call(
+            Expr {
+                kind: ExprKind::Path(QPath::Resolved(None, path)),
+                ..
+            },
+            [arg],
+        ) => {
+            if match_def_path(cx, path.res.opt_def_id()?, &paths::OPTION_SOME) {
+                Some(arg)
+            } else {
+                None
+            }
+        },
+        ExprKind::Block(
+            Block {
+                stmts: [],
+                expr: Some(expr),
+                ..
+            },
+            _,
+        ) => get_some_expr(cx, expr),
+        _ => None,
+    }
+}
+
+fn is_none_expr(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> bool {
+    match expr.kind {
+        ExprKind::Path(QPath::Resolved(None, path)) => path
+            .res
+            .opt_def_id()
+            .map_or(false, |id| match_def_path(cx, id, &paths::OPTION_NONE)),
+        ExprKind::Block(
+            Block {
+                stmts: [],
+                expr: Some(expr),
+                ..
+            },
+            _,
+        ) => is_none_expr(cx, expr),
+        _ => false,
+    }
+}
+
+fn is_none_arm(cx: &LateContext<'tcx>, arm: &'tcx Arm<'_>) -> bool {
+    if arm.guard.is_some() {
+        return false;
+    }
+    match &arm.pat.kind {
+        PatKind::Wild => true,
+        PatKind::Path(QPath::Resolved(None, path)) => path
+            .res
+            .opt_def_id()
+            .map_or(false, |id| match_def_path(cx, id, &paths::OPTION_NONE)),
+        _ => false,
+    }
+}
+
+fn get_some_binding(cx: &LateContext<'tcx>, arm: &'tcx Arm<'_>) -> Option<SomeBinding> {
+    if_chain! {
+        if arm.guard.is_none();
+        if let PatKind::TupleStruct(QPath::Resolved(None, path), [some_pat], _)  = arm.pat.kind;
+        if let Some(id) = path.res.opt_def_id();
+        if match_def_path(cx, id, &paths::OPTION_SOME);
+        then {
+            match some_pat.kind {
```

Comment:
Maybe you can skip looking at the type of pattern here. You can just take the pattern and paste it into the closure argument. Ex: `Some((x, _)) => ..` -> `map(|(x, _)| ..`. The lint will be more flexible this way.

Reply:
Not quite as simply, it would need to be irrefutable and work in the context of a closure. I'll do something with this.

Reply:
If the `match` has only one arm for `Some` (and no `_` arm), then you can be sure that the pattern is irrefutable.